### ISSUE TITLE
Finetuning Bert using ensemble methods

### ISSUE THREAD

**Issue by Riyush** [minalee-research/cs257-students#9]
#task
#method

The BERT model provides some of the most up to date word embeddings that deeply capture relationships within the data. Since it uses a transformer model to understand relationships, the task of fine-tuning BERT requires innovative ways of manipulating computed gradients for a specific corpus. Gradient Surgery is the process of minimizes destructive interference between conflicting gradients by projecting a task’s gradient onto the normal plane of another task’s gradient. Essentially, it ensures that gradient descent doesn't needlessly compound its effect on other nonorthogonal gradients leading to the vanishing gradient problem. Another interesting approach was the use of 2 datasets for finetuning. The second corpus was deliberately chosen to be different than the target corpus so as to minimise interference and contradictions in word embeddings. The second corpus would then be used to negate the vanishing gradient problem by providing a new set of orthogonal relationships to understand. This effectively prolongs the gradient of the main text for as long as needed. In the context of the sentiment classification task, the researches used the Stanford Sentiment Treebank Dataset and the CFIMBD dataset of movie reviews.  Overall, the paper tested a bunch of other fine-tuning methods and reported the comparative results below. What's interesting is the evaluations methods which utilize classification accuracy relative to computational cost. They try to measure the tradeoff between classification accuracy and computational cost. This compromise allows certain methods to show higher performance like Multitask fine-tuning and Cosine-Similarity based finetuning in specific contexts.

<img width="919" alt="Image" src="https://github.com/user-attachments/assets/07f11993-f90e-4f21-b277-7fa6e58fda9f" />
